### PR TITLE
Added writeValueObject.

### PR DESCRIPTION
**Pull request by evert**
cc: @staabm
needs tests

### CODE REVIEW

File: lib/Service.php [hunk 1]
```diff
@@ -233,6 +233,32 @@ function mapValueObject($elementName, $className) {
...
+     * @param string $contextUri
+     * @return void
+     */
+    function writeValueObject(object $object, $contextUri = null) {
+
+        if (!isset($this->classMap[get_class($object)]))
+            throw new \InvalidArgumentException('"'. get_class($object) .'" is not a registered value object class. Register your class with either mapValueObject or $classMap.');
+        }
+        return $this->write(
+            $this->classMap[get_class($object)],
```
Comment (staabm) [inline lib/Service.php]: Write req. 1st arg to be a clark notation, not a callabe

File: lib/Service.php [hunk 2]
```diff
@@ -233,6 +233,32 @@ function mapValueObject($elementName, $className) {
...
+     * intended specifically to serialize a Value Object into an XML document.
+     *
+     * The ValueObject must have been previously registered using either
+     * mapValueObject() or the $classMap property.
+     *
+     * @param object $object
+     * @param string $contextUri
+     * @return void
+     */
+    function writeValueObject(object $object, $contextUri = null) {
```
Comment (staabm) [inline lib/Service.php]: Is `object` a valid typehint?

**Comment by staabm** [comment]
Thx for picking it up. I like the pace we iterate on this lib. Thx for doing such a great job :100:

File: lib/Service.php (hunk 2 above)
Comment (evert) [inline lib/Service.php]: it wasn't =( sadly

**Comment by evert** [comment]
Thanks for all the web points!

File: tests/Sabre/Xml/ServiceTest.php [hunk 3]
```diff
@@ -231,10 +231,20 @@ function testMapValueObject() {
 
         $this->assertEquals($expected, $order);
 
-        $writtenXml = $orderService->write('{http://sabredav.org/ns}order', $order);
+        $writtenXml = $orderService->writeValueObject($order);
```
Comment (staabm) [inline tests/Sabre/Xml/ServiceTest.php]: Shouldn't we test both api calls and dont assume impl details (that those are equivalent). We should assert those are equivalent instead

File: lib/Service.php [hunk 4]
```diff
@@ -230,6 +230,33 @@ function mapValueObject($elementName, $className) {
         $this->classMap[$className] = function(Writer $writer, $valueObject) use ($namespace) {
             return \Sabre\Xml\Serializer\valueObject($writer, $valueObject, $namespace);
         };
+        $this->valueObjectMap[$className] = $elementName;
```
Comment (staabm) [inline lib/Service.php]: Note related to this PR: do we need to make sure (or help preventing users people shooting in there feet)?

- check whether a classMapping already exists before mapping the value object
- check whether the elementMapping already exists before kaüping the value object

Comment (evert) [inline lib/Service.php]: Idk! I think overwriting them is a legitimate use-case as well :/

Comment (staabm) [inline lib/Service.php]: Ok, then we should add tests for the case

Comment (evert) [inline lib/Service.php]: Or maybe not. Overwriting *inside* of Writer or Reader makes a ton of sense, but maybe not in Service. Could be good to throw an exception there

Comment (staabm) [inline lib/Service.php]: Thats what I meant ..

Comment (evert) [inline lib/Service.php]: Sorry I'm flip-flopping again. I think we need to just let people overwrite these if they want to

Comment (evert) [inline lib/Service.php]: The use-case is if an application provides a service, and a module alters the standard behavior the application provides. I don't think that that's entirely far-fetched.

Comment (staabm) [inline lib/Service.php]: :-)

**Pull request merged**